### PR TITLE
Fix failing defender functional tests

### LINKED ISSUE #1055: Defender functional tests fail due to timing/cache issues
# 🐛 Summary #

Several of the Defender policy tests fail despite the same tests being run manually succeeding.  The issue appears to be a timing/caching issue where the test changes values, but those changes are not reflected when assessed immediately following the change.  This seems particularly consistent in GCC high tenant, but it's not exclusively.
Recommendation is to change tests from "live" runs to cached runs.

## To reproduce ##

1. Setup and run the Defender standard functional tests against GCC high test tenant.
1. Watch for errors in policy groups 1 and 6.

## Expected behavior ##

All tests should have passed, but several failed.

## Any helpful log output or screenshots ##

Failure of MS.DEFENDER.1.2v1 tests despite the correct settings being enabled to pass/fail as expected
Screenshot 2024-04-03 at 11 43 47 AM

Failure of MS.DEFENDER.6.1v1 despite auditing being enabled prior to test.
Screenshot 2024-04-03 at 11 42 45 AM

### PR DESCRIPTION
**Pull request by schrolla**
#  #
  

## 🗣 Description ##

Several of the Defender functional tests were failing due to issues with change timing and precondition tenant configuration commands.  This update:
 - Adds the `-ErrorAction Continue` flag to all test PreCondition calls to `Enable-EOPProtectionPolicyRule` to prevent test failures if said policy rule is already enabled.
 - Adds a filter to test preconditions that pull a user from the tenant Entra ID directory to ensure said user information is for a user with an email address in the tenant (not a guest)
 - Changes the MS.DEFENDER.6.2v1 tests to use cached settings for testing to prevent failures due to delays in effective changes for audit settings and so as not to disable auditing in test tenants.

Additional issues with parallel execution of different Defender test plans in the tenant were also identified during testing, but is an issue with test execution control which is outside the scope of this PR.  A new issue (#1278) has been created to address this at a future date.

Note that this PR currently contains scaffolding for testing the branch which is NOT meant to be included in the final PR.  Specifically, the new Defender functional test workflow and modifications to the CallProducts.ps1 script.  These commits will be removed prior to final merge, but are included here for reviewers to view and use as needed.  Only testplan file commits should be included in the final PR before merge.

## 💭 Motivation and context ##

The changes included improve the stability of the Defender functional tests so they do not fail regularly causing lack of attention to failed nightly test results.

Closes #1055

## 🧪 Testing ##

The functional tests can be tested in two ways.  One, via manual execution of the Defender functional tests using the standard functional testing approach outlined in the [Testing README](https://github.com/cisagov/ScubaGear/blob/main/Testing/Readme.md).  To reproduce the conditions of the workflow run most closely, set `$ErrorActionPreference = 'Stop'` in your PowerShell environment before running the Defender testplans.

The other method is to use the new workflow `Defender Functional Tests (temp)`.  Previous successful run can be seen [here](https://github.com/cisagov/ScubaGear/actions/runs/10442634532).  Note:  A re-run of 2 of the tenant jobs was necessary due to the previous parallelism issue mentioned above.  However, running this on the branch either requires re-running the existing job shown or triggering a push to the branch.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] Changes are sized such that they do not touch excessive number of files. (unchecked until test scaffolding removed)
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [x] Related issues these changes resolve are linked preferably via [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] All relevant type-of-change labels added.
- [x] All relevant project fields are set.
- [x] All relevant repo and/or project documentation updated to reflect these changes.
- [x] Unit tests added/updated to cover PowerShell and Rego changes.
- [x] Functional tests added/updated to cover PowerShell and Rego changes.
- [x] All relevant functional tests passed.
- [x] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

- [x] PR passed smoke test check.
- [x] Feature branch has been rebased against changes from parent branch, as needed

  Use `Rebase branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
- [x] Resolved all merge conflicts on branch
- [ ] Notified merge coordinator that PR is ready for merge via comment mention

## ✅ Post-merge checklist ##

- [ ] Feature branch deleted after merge to clean up repository.
- [ ] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

### CODE REVIEW

**Review by Sloane4 — APPROVED** [review]

File: Testing/Functional/Products/TestPlans/defender.g5.testplan.yaml [hunk 1]
```diff
@@ -329,8 +361,13 @@ TestPlan:
...
-          - Command: 'Set-ATPProtectionPolicyRule -Identity "Strict Preset Security Policy" -SentTo (Get-MailUser)[0].UserPrincipalName -SentToMemberOf (Get-UnifiedGroup)[0].alias -RecipientDomainIs badpeople.r.us -Confirm:$false'
+          - Command: |
+                      $Domain = (@(Get-MgBetaDomain | Where-Object {$_.IsDefault -eq $true})[0]).Id
+                      $Domain = '@' + $Domain
+                      $User = (Get-MgBetaUser -ConsistencyLevel eventual -Count userCount -Filter "endsWith(UserPrincipalName, '$Domain')")[0]
+                      $Group = (Get-UnifiedGroup)[0].alias
+                      Set-ATPProtectionPolicyRule -Identity "Standard Preset Security Policy" -SentTo $User.UserPrincipalName -SentToMemberOf $Group -RecipientDomainIs badpeople.r.us -Confirm:$false
+                      Set-ATPProtectionPolicyRule -Identity "Strict Preset Security Policy" -SentTo $User.UserPrincipalName -SentToMemberOf $Group -RecipientDomainIs badpeople.r.us -Confirm:$false
```
Comment (james-garriss) [inline Testing/Functional/Products/TestPlans/defender.g5.testplan.yaml:370]: Seems like a good place for a comment to describe what you're doing.  Same for similar sections below.

File: Testing/Functional/Products/Tests/CallProductTests.ps1 [hunk 2]
```diff
@@ -7,23 +7,49 @@
...
-$params["Thumbprint"] = $thumbprint
+$Params["Thumbprint"] = $Thumbprint
 
 # Create an array of test containers
-$testContainers = @()
-$testContainers += New-PesterContainer -Path $testScriptDir -Data $params
+$TestContainers = @()
+$TestContainers += New-PesterContainer -Path $TestScriptDir -Data $Params
+
+$PesterConfig = @{
```
Comment (james-garriss) [inline Testing/Functional/Products/Tests/CallProductTests.ps1:38]: I believe that CallProductTests is called by a lot of things.  Were you able to test *all* of these calls to make sure these changes didn't impact anything else?

File: .github/workflows/test_defender_function.yaml [hunk 3]
```diff
@@ -0,0 +1,203 @@
+# Purpose: Run the product functional tests every night
+# The testing params are stored as a string in a CSV-like format.
+# Pipes separate products, and commas separate key/value pairs.
+# key1=value1,key2=value2|key3=value3,key4=value4...
+# Test version for Defender only - burn after testing
+
+name: Defender Functional Tests (temp)
```
Comment (james-garriss) [inline .github/workflows/test_defender_function.yaml:7]: Minor detail:  In order to delete a temp workflow, you have to manually delete each individual run.  Severe GitHub limitation.  Don't keep this around too long.  :-)

**Review by james-garriss — APPROVED** [review]
Good work.

**Pull request merged**